### ISSUE TITLE
Cannot easily mock out sleep function in order to simulate time in tests

### ISSUE THREAD

**Issue by asqui** [jd/tenacity#228]
Given the level of flexibility of this library, and a slightly elaborate context which I am using it in, it is desirable to be able to write an integration test where the concepts of 'time' (i.e. `time.monotonic`) and 'sleep' (i.e. `tenacity.nap.sleep`) are mocked out so that we can simulate in full fidelity how a retry scenario plays out, without having to actually sleep.

Using the [testfixtures](https://github.com/Simplistix/testfixtures) library and `mock.patch` I'm mocking out 'time' with an object that always returns the same 'virtual' time, and 'sleep' with a function that advances this 'virtual' time.

This approach works pretty well, save for one complication...

Since the signature of `BaseRetrying.__init__()` defaults the `sleep` argument to `tenacity.nap.sleep` this default is bound at import time, making it hard to mock out the sleep function after the fact.

Options I can think of:
1. Reach in to the method signature and mock out the default value - seems a bit messy.
2. Pass in a sleep function in the code under test, then mock this out in the test - I don't like this because it involves modifying production code simply for the sake of the test, and also makes this mocking strategy less portable if I want to use it in other tests (because now I need to mock out sleep at my callsite, rather than inside tenacity).
3. Make a change to the signature of `BaseRetrying.__init__()` to default `sleep` to `None`, and then default this to `nap.sleep` inside the method: `self.sleep = sleep or nap.sleep`.

Option 3 seems the neatest, because now I can just mock out `tenacity.nap.sleep` which is the most intuitive approach, but it does involve a change to the library which is why I wanted to solicit input first before preparing a PR.

Thoughts?

**Comment by dtantsur** [comment]
This issue blocks transition of [ironic](https://opendev.org/openstack/ironic) from retrying to tenacity.

I'd vote for option 3. Since `None` is not a valid sleep value, it's fine to change the default to it IMO.

**Comment by asqui** [comment]
Option 3 is unfortunately not sufficient for functions decorated with `@retry`.

When using the `@retry` decorator, this will construct a `Retrying` instance immediately, which binds `Retrying.sleep` for that instance to `nap.sleep` at import-time, and precludes any further mocking after that point.

I think we can solve both problems with a change at a lower level. Replacing the existing `nap.sleep = time.sleep` variable with a function definition `def sleep(seconds): time.sleep(seconds)` will allow patching of `nap.time.sleep` in tests.

This is essentially the same as the workaround employed here (as discussed in #25):
https://github.com/openstack/cinder/blob/307fccea4cc9c6496334b0fe137206cb48499bd5/cinder/utils.py#L56

**Comment by jd** [comment]
Why can't you mock `myfunction.retry.sleep` then?

**Comment by asqui** [comment]
It's true, when using the decorator approach you can then mock `myfunction.retry.sleep`.

I figured if we're going to make code changes to enable mocking in one of the two cases it would be nice to do it in a way that accommodates both with the same approach.

With the change in PR #236 you can use the same mocking approach (replacing `tenacity.nap.time.sleep`) regardless of exactly how the code-under-test is using Tenacity.

I think this makes writing tests:
* simpler (you only one single approach that will always work with Tenacity)
* more intuitive (you don't need to understand the fact that the decorator exposes a `retrying` member on the decorated function; you can just find `nap.sleep` quite quickly -- at least this is where I first ended up when trying to work out how to write a test for code using Tenacity.)
* more stable (tests don't need to change mocking approach if code-under-test is refactored to change from/to using the decorator)

**Comment by dtantsur** [comment]
> Why can't you mock myfunction.retry.sleep then?

Sometimes myfunction is hidden inside of something (e.g. another function, which is not uncommon for ironic).

**Comment by asqui** [comment]
Just picking up this thread again -- are we happy with this approach?
If so, I can fix up PR #236 (satisfy pylint, and switch from `mock.patch` to pytest `monkeypatch`) and rebase it ready to merge.

**Comment by jd** [comment]
TBH I'm not really excited about that change. You could also fix that yourself easily by subclassing the `Retrying` class with your own default `sleep` in `__init__` and use that through your code.

```python
def mysleep(s):
    return time.sleep(s)

class MyRetrying(tenacity.Retrying):
    def __init__(self, *args, sleep=mysleep, **kwargs):
         super().MyRetrying(*args, sleep=sleep, **kwargs)

how_to_retry = MyRetrying(…)


# Later
@how_to_retry.wraps
def myfunc_retrying():
    …
```

(untested code but you get the idea)

**Comment by dtantsur** [comment]
@jd you mean that literally everyone who is serious about their project (and thus writes unit tests) has to subclass Retrying? Forking tenacity would be less work overall..

**Comment by jd** [comment]
I don't think insulting the few thousands of tenacity users by implying they are not serious about their project as they can't write unit tests is a good approach.

Like many, I use tenacity and I write unit tests. I don't test tenacity itself and it's retrying behavior since that is tested by tenacity itself (I hope so at least). If this is what you want to do, I provided a pretty simple solution with 5 lines of code.

If you think there are **serious** issues about testing code with tenacity, I'm happy to hear them.

But if from now your plan is to bully your way in, this is not going to work.

**Comment by asqui** [comment]
@jd You said you're "not really excited about the change". It's definitely true that you can work around this already in various ways, but I just think it would be better if you didn't need to.

In my particular scenario I'm trying to standardise a myriad of retry approaches in a large codebase to all use tenacity, and convince other developers that this is worthwhile. The need to apply workarounds like this in order to be able to test is a barrier to adoption of the library by other developers.

Forking the library and applying the patch I'm suggesting is also an approach, but again carries more overhead than having this in the library as standard, and available to all users of tenacity without each needing to visit this hurdle and come up with a mitigation approach. (e.g. https://github.com/openstack/cinder/blob/307fccea4cc9c6496334b0fe137206cb48499bd5/cinder/utils.py#L56 as discussed earlier in the comment trail)

So to summarise I think there are definite advantages to applying this change to the library, and I can see no downsides.

What are the _disadvantages_ of this change which are making you "not really excited" about making it?

**Comment by dtantsur** [comment]
@jd I think you're taking critic about your projects too personally. And please don't refer to users that I allegedly insulted - there was nothing like that, and using loud words like "bully" for a minor disagreement is strange at best. Given the tone of the discussion, I'd prefer to leave this topic.

**Comment by jd** [comment]
> What are the disadvantages of this change which are making you "not really excited" about making it?

That's an excellent question to ask. I don't think there is much. I'm just trying to be really suspicious about changes that are not useful and might be a pain to maintain in the long run. That's part of my job here. 😄

**Comment by asqui** [comment]
Hmmm, so just to be clear, the change being talked about here is to change the definition of `nap.sleep` from `sleep = time.sleep` to `def sleep(seconds): time.sleep(seconds)` in order to facilitate easier mocking, in a single place, that works for a variety of usage scenarios.

It's a pretty minor change, with a major benefit for end-to-end testability of code using the library.

Is there a specific aspect of this change that you think might be a pain to maintain in the long run? If not, then I'm failing to see the reasons behind not wanting to merge such a change.

**Comment by jd** [comment]
No, I think it should be fine.

**Comment by asqui** [comment]
Ok great! I'll tidy up the PR #236 and get it ready to merge.

**Comment by asqui** [comment]
Do you have a reproducer for the patch that didn't work for you?

**Comment by ivanthewebber** [comment]
Mocking `'tenacity.nap.sleep'` wouldn't have side effects but doesn't work because wait/sleep references were already set/accessed by imports and decorators.

Mocking `'time.sleep'` or equivalently `'tenacity.nap.time.sleep'` does skip the sleep but can disrupt tests that need to use sleep or other libraries (e.g. pytest-xdist, Trio). 

What does work is if you mock `'tenacity.nap.time'`:
```python
@pytest.fixture(scope='session', autouse=True)
def disable_tenacity_sleep():
    with mock.patch('tenacity.nap.time', mock.MagicMock()):
        import time
        assert not isinstance(time, mock.MagicMock)
        from tenacity.nap import time
        assert isinstance(time, mock.MagicMock)
        yield
```

What might be slightly more ergonomic is if you split nap.sleep into two functions, one that can be mocked and other that references the potential mock. This is like using a pointer in other languages:
```python
def mockable_sleep(seconds: float) -> None:
    """
    Sleep strategy that delays execution for a given number of seconds.

    This is the default strategy, and may be the mock target for unit testing.
    """
    time.sleep(seconds)

def sleep(seconds: float) -> None:
    """
    Sleep strategy that delays execution for a given number of seconds.

    This is the default strategy, and may be mocked out for unit testing by mocking tenacity.nap.mockable_sleep.
    """
    mockable_sleep(seconds)
```

Update: I'm actually finding some tests still run slow, but are way faster with wait_none instead of normal waits, so this might not be working as intended still.

**Comment by ivanthewebber** [comment]
Could the same or similar be done for async sleep?

**Comment by asqui** [comment]
> Mocking `'time.sleep'` or equivalently `'tenacity.nap.time.sleep'` does skip the sleep but can disrupt tests that need to use sleep or other libraries (e.g. pytest-xdist, Trio).

I don't think this is true. Mocking `'tenacity.nap.time.sleep'` is not actually equivalent to mocking `'time.sleep'` (at least as far as the behaviour of `unittest.mock.patch` is concerned).

See: https://docs.python.org/3/library/unittest.mock.html#where-to-patch

**Comment by ivanthewebber** [comment]
@asqui easy to test:
```
$ poetry run python
Python 3.14.2 (main, Dec 18 2025, 12:15:05) [Clang 17.0.0 (clang-1700.0.13.5)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Cmd click to launch VS Code Native REPL
>>> from unittest import mock
>>> with mock.patch("tenacity.nap.time.sleep", mock.MagicMock()):
...     import time
...     print(type(time.sleep))
... 
<class 'unittest.mock.MagicMock'>
```

**Comment by asqui** [comment]
Yeah you're right, I'm talking nonsense. Apologies.

Why don't you mock `tenacity.nap.sleep`? It was added for this purpose.

(Unfortunately I think some of the guidance further up in the issue comments saying to mock `tenacity.nap.time.sleep` is wrong -- for the reasons you outline.)
